### PR TITLE
Metrics and diagnostics service

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**
## Usage and product changes

We introduce a metrics and diagnostics service. The web endpoint is bound at port 4104 by default and exposes metrics formatted for prometheus (`http://localhost:4104/metrics?format=prometheus`) or as JSON (`http://localhost:4104/metrics?format=JSON`).

Example metrics:
```
# TypeDB version: TypeDB Core 2.26.6
# Time zone: Europe/London
# Java version: Azul Systems, Inc. 11.0.15
# Platform: Mac OS X aarch64 13.6.3

# TYPE typedb_attempted_requests_total counter
typedb_attempted_requests_total{kind="CONNECTION_OPEN"} 625
typedb_attempted_requests_total{kind="SERVERS_ALL"} 625
typedb_attempted_requests_total{kind="USER_MANAGEMENT"} 0
typedb_attempted_requests_total{kind="USER"} 0
typedb_attempted_requests_total{kind="DATABASE_MANAGEMENT"} 2606
typedb_attempted_requests_total{kind="DATABASE"} 252
typedb_attempted_requests_total{kind="SESSION"} 1207
typedb_attempted_requests_total{kind="TRANSACTION"} 2835

# TYPE typedb_successful_requests_total counter
typedb_successful_requests_total{kind="CONNECTION_OPEN"} 625
typedb_successful_requests_total{kind="SERVERS_ALL"} 625
typedb_successful_requests_total{kind="USER_MANAGEMENT"} 0
typedb_successful_requests_total{kind="USER"} 0
typedb_successful_requests_total{kind="DATABASE_MANAGEMENT"} 2606
typedb_successful_requests_total{kind="DATABASE"} 252
typedb_successful_requests_total{kind="SESSION"} 1207
typedb_successful_requests_total{kind="TRANSACTION"} 2755

# TYPE typedb_current_count gauge
typedb_current_count{kind="DATABASES"} 2
typedb_current_count{kind="SESSIONS"} 1
typedb_current_count{kind="TRANSACTIONS"} 1
typedb_current_count{kind="USERS"} 0

# TYPE typedb_error_total counter
typedb_error_total{code="TXN23"} 2
typedb_error_total{code="QRY22"} 4
typedb_error_total{code="TXN19"} 2
typedb_error_total{code="THR10"} 19
```

We also extract the metrics reporter into a separate module within the diagnostics package and redirect it to the dedicated diagnostics service.

### CODE REVIEW

File: common/diagnostics/Metrics.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    public void requestAttempt(Metrics.NetworkRequests.Kind kind) {
+        requests.attempt(kind);
+    }
+
+    public void requestSuccess(Metrics.NetworkRequests.Kind kind) {
+        requests.success(kind);
+    }
+
+    public void setCurrentCount(UsageCounters.Kind kind, long value) {
```
Comment (flyingsilverfin) [inline common/diagnostics/Metrics.java:49]: the other methods are called `<type>Attempt/success/etc.` and this one is mismatched between method name and counter type, is that intended? maybe `setUsageCount`?

File: common/diagnostics/Metrics.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            return system;
+        }
+
+        String formatPrometheus() {
+            return "# TypeDB version: " + name + " " + version + "\n" +
+                    // no serverID, that's for reporting only
+                    "# Time zone: " + TimeZone.getDefault() .getID() + "\n" +
+                    "# Java version: " + System.getProperty("java.vendor") + " " + System.getProperty("java.version") + "\n" +
+                    "# Platform: " + System.getProperty("os.name") + " " + System.getProperty("os.arch") + " " + System.getProperty("os.version") + "\n";
+        }
```
Comment (flyingsilverfin) [inline common/diagnostics/Metrics.java:101]: Is it possible for these to error at any point? What will happen if so? (eg. maybe we should wrap in a try/catch so the server doesn't happe to crash in some envs?)

File: common/diagnostics/MonitoringServer.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            ServerBootstrap bootstrap = new ServerBootstrap();
+            bootstrap.group(group).channel(NioServerSocketChannel.class).childHandler(new MetricsInitializer(sslContext, middleware));
+            Channel channel = bootstrap.bind(scrapePort).sync().channel();
+            channel.closeFuture().sync();
+        } catch (InterruptedException ignored) {
+            if (LOG.isDebugEnabled()) LOG.debug("Monitoring server interrupted.");
+            // do nothing
+        } finally {
+            group.shutdownGracefully();
+        }
```
Comment (flyingsilverfin) [inline common/diagnostics/MonitoringServer.java:86]: Did you validate that shutdown is still quick/works OK when this is running?

File: common/diagnostics/MonitoringServer.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        @Override
+        public void channelReadComplete(ChannelHandlerContext ctx) {
+            ctx.flush();
+        }
+
+        @Override
+        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Failed to respond to a metrics request:\n{}\n", Arrays.toString(cause.getStackTrace()));
```
Comment (flyingsilverfin) [inline common/diagnostics/MonitoringServer.java:150]: I believe the LOGger can take an exception as an argument and will handle formatting and stuff for you, check some other places we log errors

File: server/parameters/config.yml [hunk 5]
```diff
@@ -64,7 +64,11 @@ log:
 
 diagnostics:
   reporting:
+    errors: true
+    statistics: true
+  monitoring:
```
Comment (flyingsilverfin) [inline server/parameters/config.yml:69]: Should we disable monitoring by default? Or leave it on?

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: common/diagnostics/Metrics.java (hunk 2 above)
Comment (dmitrii-ubskii) [reply common/diagnostics/Metrics.java:101]: Empty `getProperty()` is not an error, and `TimeZone.getDefault()` is never `null` (it's set to GMT as last resort), so `.getID()` should not be able to error either,

File: common/diagnostics/Metrics.java (hunk 1 above)
Comment (dmitrii-ubskii) [reply common/diagnostics/Metrics.java:49]: `setCurrentCount(DATABASES, count)` reads better on callsite. Renamed `UsageCounters` to `CurrentCounts`.

File: common/diagnostics/Diagnostics.java [hunk 6]
```diff
@@ -18,63 +18,146 @@
...
+                return;
+            }
+
+            initSentry(serverID, distributionName, version, errorReportingEnable, errorReportingURI);
+
+            Metrics metrics = new Metrics(serverID, distributionName, version);
+            StatisticReporter statisticReporter = initStatisticReporter(statisticsReportingEnable, statisticsReportingURI, metrics);
+            MonitoringServer monitoringServer = initMonitoringServer(monitoringEnable, monitoringPort, metrics);
+
+            diagnostics = new Core(metrics, statisticReporter, monitoringServer);
```
Comment (flyingsilverfin) [inline common/diagnostics/Diagnostics.java:97]: So these are just `null`able? Where do we do the null check to do nothing if the reporter/server is null?

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**